src/GdbCommand.cc

```diff
@@ -31,11 +31,15 @@ static std::vector<ReplayTimeline::Mark> back_stack;
 static ReplayTimeline::Mark current_history_cp;
 static std::vector<ReplayTimeline::Mark> forward_stack;
 static SimpleGdbCommand rr_history_push(
-    "rr-history-push", [](GdbServer& gdb_server, Task*, const vector<string>&) {
+    "rr-history-push", [](GdbServer& gdb_server, Task* t, const vector<string>&) {
       if (current_history_cp) {
         back_stack.push_back(current_history_cp);
       }
-      current_history_cp = gdb_server.get_timeline().mark();
+      if (!t->session().is_replaying()) {
```

Comment:
I think it's a bit subjective. If you assume diversion are short-lived to perhaps evaluate a quick value, then yes you'd want no side effects. If longer running diversion are more typical then it make sense to consider consider that a 'navigation', clear the forward stack, and set the current mark as undefined.

But I don't use diversion with RR so I don't really know. I'll update the patch to have no side effects for now and I'm sure we will notice if that behavior is not good and we can change it.